# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

In [2]:
os.getcwd()

'/Users/Cyril'

In [3]:
path="data"
os.chdir(path) 

In [4]:
os.getcwd()

'/Users/Cyril/data'

In [5]:
dir_list = os.listdir() 
print("Files and directories in  current working directory :")  
print(dir_list) 

Files and directories in  current working directory :
['step2_output.csv', 'step3_output.csv']


**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [6]:
df = pd.read_csv('step3_output.csv')
df.head()

Name    state  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort   Alaska         3939           2500        250      1   
1  Eaglecrest Ski Area   Alaska         2600           1540       1200      0   
2     Hilltop Ski Area   Alaska         2090            294       1796      0   
3     Arizona Snowbowl  Arizona        11500           2300       9200      0   
4  Sunrise Park Resort  Arizona        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  ...  SkiableTerrain_ac  \
0        0.0          0          2     2  ...             1610.0   
1        0.0          0          0     0  ...              640.0   
2        0.0          0          0     0  ...               30.0   
3        0.0          1          0     2  ...              777.0   
4        0.0          0          1     2  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         0  
1          53.0               90.0             0.0         0  
2          34.0              152.0            30.0         0  
3          89.0              122.0             0.0         1  
4          78.0              104.0            80.0         1  

[5 rows x 27 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [7]:
dummy = pd.get_dummies(df['state'])

df = pd.concat([df.drop(columns=['state']), dummy], axis=1)

df.head()

Name  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort         3939           2500        250      1   
1  Eaglecrest Ski Area         2600           1540       1200      0   
2     Hilltop Ski Area         2090            294       1796      0   
3     Arizona Snowbowl        11500           2300       9200      0   
4  Sunrise Park Resort        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  triple  ...  Rhode Island  \
0        0.0          0          2     2       0  ...             0   
1        0.0          0          0     0       0  ...             0   
2        0.0          0          0     0       1  ...             0   
3        0.0          1          0     2       2  ...             0   
4        0.0          0          1     2       3  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 61 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [8]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [9]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to raveradddf and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [10]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [12]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test

y_pred = model.predict(X_test)


## Review Model Outcomes — Iterate over additional models as needed

In [13]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [15]:
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error

print(explained_variance_score(y_test, y_pred))

print(mean_absolute_error(y_test, y_pred))



0.9307574966933976
5.1615904116948155


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [16]:
print(lm.intercept_)

12.126362839002454


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [17]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])


Coefficient
summit_elev           0.004945
vertical_drop         0.006207
base_elev             0.005727
trams                 0.187369
fastEight             0.468926
fastSixes             0.676102
fastQuads             0.013065
quad                  0.678083
triple                0.803751
double                0.152955
surface               0.831487
total_chairs          0.395692
Runs                  0.026796
TerrainParks          0.092037
LongestRun_mi         0.409402
SkiableTerrain_ac     0.000094
Snow Making_ac        0.003669
daysOpenLastYear      0.009986
yearsOpen             0.003418
averageSnowfall       0.001533
AdultWeekday          0.805035
projectedDaysOpen     0.003596
NightSkiing_ac        0.007804
clusters              0.616748
Alaska                2.797090
Arizona               6.496288
California            6.612768
Colorado              5.778081
Connecticut           4.734262
Idaho                 7.263714
Illinois              1.071592
Indiana               5.158444
Iowa                  3.932845
Maine                 2.695396
Maryland             10.262079
Massachusetts         8.652273
Michigan              0.149483
Minnesota             2.851344
Missouri              3.487126
Montana               9.955593
Nevada                1.008143
New Hampshire         6.834490
New Jersey            1.420340
New Mexico            9.829089
New York              0.425349
North Carolina       13.043948
Ohio                  6.067552
Oregon                5.822722
Pennsylvania          3.334690
Rhode Island          4.244385
South Dakota          7.094605
Tennessee            19.016762
Utah                  1.868684
Vermont               1.992299
Virginia              7.043905
Washington            3.387390
West Virginia        10.372738
Wisconsin             1.139279
Wyoming               6.722826

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [18]:
df = pd.read_csv('step3_output.csv')


# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','state','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to raveradddf and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)


#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)



y_pred = model.predict(X_test)


from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error

print(explained_variance_score(y_test, y_pred))

print(mean_absolute_error(y_test, y_pred))



print(lm.intercept_)


pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])


0.9236919464675647
5.523311392028669
14.042764960670667


Coefficient
summit_elev           0.003578
vertical_drop         0.005241
base_elev             0.003633
trams                 0.663925
fastEight             0.537374
fastSixes             0.467624
fastQuads             0.166900
quad                  0.990351
triple                0.692756
double                0.012268
surface               0.828118
total_chairs          0.300482
Runs                  0.035610
TerrainParks          0.111107
LongestRun_mi         0.252870
SkiableTerrain_ac     0.000203
Snow Making_ac        0.001913
daysOpenLastYear      0.027083
yearsOpen             0.002617
averageSnowfall       0.010789
AdultWeekday          0.834161
projectedDaysOpen     0.010780
NightSkiing_ac        0.005641
clusters              1.667852

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [19]:
df = pd.read_csv('step3_output.csv')


# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','state','AdultWeekend','summit_elev','base_elev'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to raveradddf and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)


#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)



y_pred = model.predict(X_test)


from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error

print(explained_variance_score(y_test, y_pred))

print(mean_absolute_error(y_test, y_pred))



print(lm.intercept_)


pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])


0.9300201909487694
5.3269338560282335
13.564363470027288


Coefficient
vertical_drop         0.001857
trams                 0.968549
fastEight             1.348402
fastSixes             0.302714
fastQuads             0.065210
quad                  0.721887
triple                0.450268
double                0.223111
surface               1.031773
total_chairs          0.534626
Runs                  0.037534
TerrainParks          0.085095
LongestRun_mi         0.236440
SkiableTerrain_ac     0.000212
Snow Making_ac        0.001536
daysOpenLastYear      0.029105
yearsOpen             0.002574
averageSnowfall       0.010469
AdultWeekday          0.833480
projectedDaysOpen     0.009594
NightSkiing_ac        0.006005
clusters              1.601825

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. |0.9307574966933976 | 5.1615904116948155|-|
| Model 2. | 0.9236919464675647|5.523311392028669 |'state'|
| Model 3. | 0.9300201909487694 |5.3269338560282335 |'state','summit_elev','base_elev'|

Model Selection:

In [ ]:
# Model 1 is selected, becase it has the higer "Explained Variance" with lower "Mean Absolute Error".